### ISSUE TITLE
Running time problem for high dimensions

### ISSUE THREAD

**Issue by YotamNitzan** [uumami/mhar#4]
Hi Mario,
Thanks for your research and this repository!

I'm trying to use the repository but encounter unreasonable running times for high dimensions. The results reported in the paper seem much better so I assume I'm missing something.

I have N+1 points in N dimensions, defining a convex polytope/hull. 
For a small N, the code seems to run perfectly. However, when using N~300, which is my real use case, the "walk" function simply doesn't temrminate.

A minimal example to repoduce my code is attached below.

```
from mhar import walk
import numpy as np
from scipy.spatial import ConvexHull
import torch

n = torch.cuda.device_count()
print(n)

N=300
pts = np.random.rand(N+1, N)
hull = ConvexHull(pts)

ai = torch.from_numpy(hull.equations[:,:-1]).to(torch.float32).cuda()
bi = torch.from_numpy(-1 * hull.equations[:,-1]).unsqueeze(1).to(torch.float32).cuda()
x_0 = torch.from_numpy(np.mean(pts, axis=0)).unsqueeze(1).to(torch.float32).cuda()

X = walk(z=5, # Padding Parameter
     ai=ai, # Inequality matrix
     bi=bi, # Inequaliy restrictions
     ae=torch.empty(0), # Equality Matrix
     be=torch.empty(0), # Equality restriction
     x_0=x_0, # Inner Point of the polytope
     T=1, # Number of iid Iterations
     device='cuda', # Device to use cpu or cuda
     warm=0, # Number of iid iterations to burn before saving samples
     seed=None, # Automatic random seed
     thinning=None # Automatic thinning factor of O(n^3)
     )
```

In the paper, Table B.5, for hypercube with n=500 and z=4000 - running time is reported to be 27.5 seconds.
The code above has been running for an hour now and not terminating.
Could you please help me understanding what I'm missing?

Thanks!

**Comment by uumami** [comment]
Are you using a GPU?
Are you running it in colab or locally?

For using a GPU in colab:https://colab.research.google.com/notebooks/gpu.ipynb

The times are reported for GPU!

In any case knowing where you are executing the code would help!

**Comment by YotamNitzan** [comment]
I'm using a GPU. I ran the attached code snippet both locally and in colab with GPU enabled.
Locally, I see modest usage on the GPU. 25% GPU utilization and 1.2GB Mem.

As you can also see in the snippet, I passed device='cuda' to the walk function as well as loaded ai,bi,x_0 to GPU.
Is the something else I'm missing?

Would you be able to confirm how long it takes you to run this code?
It would help me understand if I'm doing something wrong or that execution times are just long.

Last, the automatic thinning factor of 300^3 is obviously large and causes longer execution times. Is there anything I can do on that end?

**Comment by uumami** [comment]
Oh I see. The results from table  Table B.5 are without the thinning factor (or a thinning factor of 0) it is raw sampling. That is the reason is taking so long, because you are dropping 300^3 points. 

You can pass the exact thinning factor you want with the parameter "thinning" int the walk function. It can be any integer you want.

You can increase the parameter z until your memory is almost full that is the "sweet spot".

The general strategy is to find a z that gives you around 95% of memory usage. The algorithm takes some time to produce iid observations, but when it produces it will give you z iid points.

For colab the optimal seems to be around z=2300:
X = walk(**z=2300, # Padding Parameter**
     ai=ai, # Inequality matrix
     bi=bi, # Inequaliy restrictions
     ae=torch.empty(0), # Equality Matrix
     be=torch.empty(0), # Equality restriction
     x_0=x_0, # Inner Point of the polytope
     T=1, # Number of iid Iterations
     device='cuda', # Device to use cpu or cuda
     warm=0, # Number of iid iterations to burn before saving samples
     seed=None, # Automatic random seed
     **thinning=None # Automatic thinning factor of O(n^3)**
     )
